zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentReconstructProcessor.java

```diff
@@ -42,7 +53,134 @@ public DeploymentReconstructProcessor(
   @Override
   public void processRecord(final TypedRecord<DeploymentRecord> record) {
     final var key = keyGenerator.nextKey();
-    // TODO: Only append the event when there are no more deployments to reconstruct
-    stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
+    final var resource = findNextResource();
+    if (resource == null) {
+      stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
+      return;
+    }
+
+    final DeploymentRecord deploymentRecord;
+    if (resource.deploymentKey() != NO_DEPLOYMENT_KEY) {
+      final var allResourcesOfDeployment = findResourcesWithDeploymentKey(resource.deploymentKey());
+      deploymentRecord =
+          recreateDeploymentForResources(
+              resource.deploymentKey(), resource.tenantId(), allResourcesOfDeployment);
+    } else {
+      deploymentRecord = createNewDeploymentForResource(resource);
+    }
+    stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED, deploymentRecord);
+  }
+
+  private Resource findNextResource() {
+    final var foundProcess = new MutableReference<PersistedProcess>();
+    processState.forEachProcess(
+        null, // TODO: Continue where we left off
+        process -> {
+          final var deploymentKey = process.getDeploymentKey();
+          if (deploymentKey != NO_DEPLOYMENT_KEY
+              && deploymentState.hasStoredDeploymentRecord(deploymentKey)) {
+            return true;
+          }
+          final var copy = new PersistedProcess();
+          BufferUtil.copy(process, copy);
+          foundProcess.set(copy);
+          return false;
+        });
+
+    if (foundProcess.get() != null) {
+      new ProcessResource(foundProcess.get());
+    }
+    // TODO: Continue with formState and decisionState
+
+    return null;
+  }
+
+  private Set<Resource> findResourcesWithDeploymentKey(final long deploymentKey) {
+    // Iterate through processState, formState, and decisionState to find resources that are marked
+    // with the given deployment key.
+    final var resources = new HashSet<Resource>();
+    processState.forEachProcess(
+        null,
```

Comment:
❌ Instead of starting from the beginning we could start with the deployment key. It seems the deployment key is always generated as a first step when processing a new deployment. That means subsequent resource keys must have a higher value. By doing this we don't have to go through the entire state all the time. We start with the first process in the deployment, and we can stop as soon as we encounter a process that has a deployment key that's not equal to the provided key.

Reply:
We should add a test case to ensure this behavior remains 🙂

Reply:
Processes are stored by tenant first so there's no way to start iteration at a certain deployment key.

Reply:
Because in the meantime a new deployment for a tenant with a lower alphanumeric value occurred?

Reply:
I meant that the problem is fundamentally that you can only iterate through processes in tenant id order - so if we are searching for all processes with a given deployment key, where should we start searching?

Reply:
That's only because you pass the deployment key to this method. You have access to a full resource which contains a tenant id. So you do know what the previous tenant is and could continue from there.

Reply:
Ah right, because a deployment can't contain resources from multiple tenants? Yeah that would work then :+1:

Reply:
I strongly believe we should do this. Because depending on the amount of processes this could be a real bottleneck performance-wise 😄

Reply:
Done in 95fb89c